auth: manual token expiration, better auxiliary token handling, improve tests

manicminer opened the pull request:
## Better Handling of Auxiliary Tokens

Adds the `AuthorizeRequest` field to `client.Client` which API clients can define for customized authoriation behavior. This is then utilized by `resourcemanager.Client` to decorate requests with both `Authorization` and `x-ms-authorization-auxiliary` headers.

Adds a helper function `SetAuthHeader()` which decorates an `*http.Request` with a bearer token, e.g. adds the `Authorization` header. This is called by `base.Client` when `AuthorizeRequest` is nil.

## Manual Token Expiration

This provides the ability to intentionally expire a cached access token, which causes a new token to be automatically acquired the next time the `Token()` or `AuxiliaryTokens()` methods are called.

Downstream implementers can test whether an `auth.Authorizer` implements `auth.CachingAuthorizer`, in case they receive a standard `auth.Authorizer`, and then call `InvalidateTokenCache()` prior to making a new request. This requires that the method is called as-needed when an action is taken that is likely to necessitate a new access token, for example with management groups which make use of token claims for access control.

Closes: #459 

Related:
- https://github.com/hashicorp/terraform-provider-azurerm/issues/12478
- https://github.com/hashicorp/terraform-provider-azurerm/issues/15894
- https://github.com/hashicorp/terraform-provider-azurerm/issues/19595
- https://github.com/hashicorp/terraform-provider-azurerm/issues/23238

manicminer commented:
Unit test results

<img width="528" alt="Screenshot 2023-09-27 at 13 04 58" src="https://github.com/hashicorp/go-azure-sdk/assets/251987/7c1020c6-fb0d-4642-a750-ee055fe1dd90">

industrialzombie commented:
Hey,

The `client` will need to be updated to use this new `SetAuthHeaders()` function when it executes a request.
Should that be included in the scope of this change?

https://github.com/hashicorp/go-azure-sdk/blob/main/sdk/client/client.go#L337

manicminer commented:
@industrialzombie Yes it will, I had thought to add that separately for additional safety downstream, but on balance it's probably fine to add it here and merge it all together 👍

industrialzombie commented:
Is it possible to get an estimate on when this might be reviewed?

manicminer commented:
@industrialzombie We'll look at this after today's release, so that we have a few days to catch any potential bugs before shipping.

industrialzombie commented:
Thanks, I appreciate the feedback!

tombuildsstuff commented on sdk/auth/cached_authorizer.go:
```diff
@@ -67,6 +68,15 @@ func (c *CachedAuthorizer) AuxiliaryTokens(ctx context.Context, req *http.Reques
 
+// ExpireTokens expires the currently cached token and auxTokens, forcing new tokens to be acquired when Token() or AuxiliaryTokens() are next called
+func (c *CachedAuthorizer) ExpireTokens() error {
+	c.token.Expiry = time.Now()
```
sanity-checking: what if `c.Token` is nil here?

manicminer replied:
thanks, good catch

tombuildsstuff commented on sdk/auth/interface.go:
```diff
@@ -13,8 +13,8 @@ import (
 // Authorizer is anything that can return an access token for authorizing API connections
 type Authorizer interface {
 	Token(ctx context.Context, request *http.Request) (*oauth2.Token, error)
-
 	AuxiliaryTokens(ctx context.Context, request *http.Request) ([]*oauth2.Token, error)
+	ExpireTokens() error
```
for this one (but also in retrospect) is it worth adding some documentation for this one? As it stands the "will do nothing" comments on each implementation will show as needed, but don't strictly explain the purpose for this one?

manicminer replied:
good shout, will add 👍

tombuildsstuff commented on sdk/auth/token.go:
```diff
@@ -4,15 +4,54 @@
 
+// SetAuthHeaders decorates a *http.Request with necessary authorization headers for Azure APIs. For more information about the vendor-specific
+// `x-ms-authorization-auxiliary` header, see https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/authenticate-multi-tenant
+func SetAuthHeaders(ctx context.Context, req *http.Request, authorizer Authorizer) error {
```
since this is specific to Resource Manager, this probably wants to go into the Resource Manager Client?

manicminer replied:
agreed, other APIs may support the concept of multiple tenants but the implementation will be different

tombuildsstuff commented on sdk/auth/token.go:
```diff
@@ -26,7 +65,11 @@ func tokenDueForRenewal(token *oauth2.Token) bool {
-	expiresWithinTenMinutes := expiry.Add(-delta).Before(now)
+
+	// Always return early if the token validity doesn't extend past the expiry delta
+	if expiry.Add(-delta).Before(now) {
```
There's a `sub`tract we can use here:

```suggestion
	if expiry.Sub(delta).Before(now) {
```

manicminer replied:
`Sub` actually does the inverse and subtracts two `Time` structs to return a `Duration`

tombuildsstuff commented on sdk/client/client.go:
```diff
@@ -327,14 +327,11 @@ func (c *Client) Execute(ctx context.Context, req *Request) (*Response, error) {
 
-	// at this point we're ready to send the HTTP Request, as such let's get the Authorization token
-	// and add that to the request
+	// Set Authorization and X-Ms-Authorization-Auxiliary headers
```
Doing this will cause issues with Storage Data Plane which is _super_ picky about the headers it supports

Perhaps Client should have a `ConfigureAuthorizerForRequest` function/method defined against the struct which the implementations can pass-in?

manicminer replied:
Makes sense, I've added this and set a custom authorization function for `ResourceManagerClient`. Where such custom function is not specified, the base client will fall back to setting a standard `Authorization` header with a bearer token (i.e. its current behavior).

tombuildsstuff commented on sdk/client/client.go:
```diff
@@ -327,14 +327,11 @@ func (c *Client) Execute(ctx context.Context, req *Request) (*Response, error) {
 
-	// at this point we're ready to send the HTTP Request, as such let's get the Authorization token
-	// and add that to the request
+	// Set Authorization and X-Ms-Authorization-Auxiliary headers
```
Doing this will cause issues with Storage Data Plane which is _super_ picky about the headers it supports

Perhaps Client should have a `ConfigureAuthorizerForRequest` function/method defined against the struct which the implementations can pass-in?

tombuildsstuff commented on sdk/auth/shared_key_authorizer.go:
```diff
@@ -64,7 +64,10 @@ func (s *SharedKeyAuthorizer) AuxiliaryTokens(_ context.Context, _ *http.Request
-// ---
+// ExpireToken has no effect with shared keys
+func (c *SharedKeyAuthorizer) ExpireTokens() error {
+	return fmt.Errorf("SharedKeyAuthorizer tokens cannot expire")
```
given the other authorizers are a noop, why not do the same here? perhaps this method wants renaming to `InvalidateTokenCache` or something to better reflect what it's doing?

manicminer replied:
I've done away with this ambiguity and added a `CachingAuthorizer` interface which can be tested for at runtime to call the `InvalidateTokenCache()` method.

tombuildsstuff reviewed (approved):
Left some comments inline but otherwise 👍

manicminer commented:
Thanks for the review @tombuildsstuff, I've reworked it to remove the anbiguity over invalidation by adding a `CachingAuthorizer` interface, and added an `AuthorizeRequest` property to the base client to allow API clients to supply a custom function for decorating the request. WDYT?

tombuildsstuff reviewed (approved):
A couple of minor comments but otherwise 👍
